internal/types/subtypes/interceptor.go

```diff
@@ -0,0 +1,193 @@
+package subtypes
+
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/sdk/pbs/controller/protooptions"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+func messageDomain(m proto.Message) string {
+	r := m.ProtoReflect()
+	fd := r.Descriptor().ParentFile()
+	if fd == nil {
+		return ""
+	}
+
+	opts, ok := fd.Options().(*descriptorpb.FileOptions)
+	if !ok {
+		return ""
+	}
+
+	domain := proto.GetExtension(opts, protooptions.E_Domain).(string)
```

Comment:
This does seem to be safe, and will result in an empty string being returned if domain is not set on the file. I don't think an error is necessary, since the only thing I can see to do if a domain is set, is potentially return from `transformRequestAttributes` if `domain == ""`